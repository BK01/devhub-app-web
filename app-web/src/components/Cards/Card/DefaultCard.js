import React from 'react';
import PropTypes from 'prop-types';
import DotDotDot from 'react-dotdotdot';
import NameSpacedImg from '../../UI/NameSpacedImg/NameSpacedImg';
import styles from './Card.module.css';
import Card from '../../../hoc/Card';
import Link from '../../UI/Link/Link';
import { ARIA_LABEL_RESOURCE } from '../../../constants/ariaLabels';
import { CARD_CONFIG } from '../../../constants/ui';

const DefaultCard = ({ title, description, resourcePath, image, resourceType, author }) => (
  <Card resourceType={resourceType} resourcePath={resourcePath} title={title} author={author}>
    <div className={styles.Body}>
      <DotDotDot clamp={CARD_CONFIG.maxDescriptionLines} className={styles.BodyDescription}>
        <p>{description}</p>
      </DotDotDot>
      <div className={styles.BodyImage}>
        {image && image !== '' ? (
          <Link to={resourcePath} aria-label={ARIA_LABEL_RESOURCE}>
            <NameSpacedImg src={image || ''} unstyled />
          </Link>
        ) : null}
      </div>
    </div>
  </Card>
);

DefaultCard.displayName = 'Default Card Component';

DefaultCard.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  resourcePath: PropTypes.string.isRequired,
  resourceType: PropTypes.string,
  image: PropTypes.string,
  author: PropTypes.string,
};

// resource type defaults to blank, because any graph ql nodes that have no resource type
// default to rendering this component
DefaultCard.defaultProps = {
  image: '',
  resourceType: '',
  author: '',
};

export default DefaultCard;
