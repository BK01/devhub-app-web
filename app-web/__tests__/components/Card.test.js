import React from 'react';
import { shallow } from 'enzyme';
import DefaultCard from '../../src/components/Cards/Card/DefaultCard';

describe('Card Component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <DefaultCard title="title" description="description" resourcePath="resourcePath" />,
    );
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders no image in the body image if image is invalid', () => {
    expect(
      wrapper
        .find('.bodyImage')
        .children()
        .exists(),
    ).toBe(false);
  });

  // flag.js appear to be breaking this test. this needs to be explored
  // it('renders an image in the body image is passed one', () => {
  //   const wrapper = shallow(
  //     <DefaultCard
  //       title="title"
  //       description="description"
  //       resourcePath="resourcePath"
  //       image="matt damon"
  //     />,
  //   );
  //   expect(
  //     wrapper
  //       .find('.bodyImage')
  //       .children()
  //       .exists(),
  //   ).toBe(true);
  // });
});
