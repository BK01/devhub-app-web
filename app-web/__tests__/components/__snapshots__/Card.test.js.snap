// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Default Card Component
    author=""
    description="description"
    image=""
    resourcePath="resourcePath"
    resourceType=""
    title="title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "author": "",
      "children": <div
        className="Body"
      >
        <Dotdotdot
          clamp={4}
          className="BodyDescription"
          tagName="div"
          truncationChar="…"
          useNativeClamp={true}
        >
          <p>
            description
          </p>
        </Dotdotdot>
        <div
          className="BodyImage"
        />
      </div>,
      "resourcePath": "resourcePath",
      "resourceType": "",
      "title": "title",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Dotdotdot
            clamp={4}
            className="BodyDescription"
            tagName="div"
            truncationChar="…"
            useNativeClamp={true}
          >
            <p>
              description
            </p>
          </Dotdotdot>,
          <div
            className="BodyImage"
          />,
        ],
        "className": "Body",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <p>
              description
            </p>,
            "clamp": 4,
            "className": "BodyDescription",
            "tagName": "div",
            "truncationChar": "…",
            "useNativeClamp": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "description",
            },
            "ref": null,
            "rendered": "description",
            "type": "p",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
            "className": "BodyImage",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "author": "",
        "children": <div
          className="Body"
        >
          <Dotdotdot
            clamp={4}
            className="BodyDescription"
            tagName="div"
            truncationChar="…"
            useNativeClamp={true}
          >
            <p>
              description
            </p>
          </Dotdotdot>
          <div
            className="BodyImage"
          />
        </div>,
        "resourcePath": "resourcePath",
        "resourceType": "",
        "title": "title",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Dotdotdot
              clamp={4}
              className="BodyDescription"
              tagName="div"
              truncationChar="…"
              useNativeClamp={true}
            >
              <p>
                description
              </p>
            </Dotdotdot>,
            <div
              className="BodyImage"
            />,
          ],
          "className": "Body",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <p>
                description
              </p>,
              "clamp": 4,
              "className": "BodyDescription",
              "tagName": "div",
              "truncationChar": "…",
              "useNativeClamp": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "description",
              },
              "ref": null,
              "rendered": "description",
              "type": "p",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": null,
              "className": "BodyImage",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
